driver/src/main/java/org/neo4j/driver/internal/bolt/pooledimpl/PooledBoltConnectionProvider.java

```diff
@@ -0,0 +1,648 @@
+                                .filter(entry -> entry.available)
+                                .count();
+                    }
+                });
+        return boltConnectionProvider.init(
+                address, routingContext, boltAgent, userAgent, connectTimeoutMillis, metricsListener);
+    }
+
+    @SuppressWarnings({"ReassignedVariable"})
+    @Override
+    public CompletionStage<BoltConnection> connect(
+            SecurityPlan securityPlan,
+            DatabaseName databaseName,
+            Supplier<CompletionStage<Map<String, Value>>> authMapStageSupplier,
+            AccessMode mode,
+            Set<String> bookmarks,
+            String impersonatedUser,
+            BoltProtocolVersion minVersion,
+            NotificationConfig notificationConfig,
+            Consumer<DatabaseName> databaseNameConsumer) {
+        synchronized (this) {
+            if (closeStage != null) {
+                return CompletableFuture.failedFuture(new IllegalStateException("Connection provider is closed."));
+            }
+        }
+
+        var acquisitionFuture = new CompletableFuture<PooledBoltConnection>();
+
+        authMapStageSupplier.get().whenComplete((authMap, authThrowable) -> {
+            if (authThrowable != null) {
+                acquisitionFuture.completeExceptionally(authThrowable);
+                return;
+            }
+
+            var beforeAcquiringOrCreatingEvent = metricsListener.createListenerEvent();
+            metricsListener.beforeAcquiringOrCreating(poolId, beforeAcquiringOrCreatingEvent);
+            acquisitionFuture.whenComplete((connection, throwable) -> {
+                throwable = FutureUtil.completionExceptionCause(throwable);
+                if (throwable != null) {
+                    if (throwable instanceof TimeoutException) {
+                        metricsListener.afterTimedOutToAcquireOrCreate(poolId);
+                    }
+                } else {
+                    metricsListener.afterAcquiredOrCreated(poolId, beforeAcquiringOrCreatingEvent);
+                }
+                metricsListener.afterAcquiringOrCreating(poolId);
+            });
+            connect(
+                    acquisitionFuture,
+                    securityPlan,
+                    databaseName,
+                    authMap,
+                    authMapStageSupplier,
+                    mode,
+                    bookmarks,
+                    impersonatedUser,
+                    minVersion,
+                    notificationConfig);
+        });
+
+        return acquisitionFuture
+                .whenComplete((ignored, throwable) -> {
+                    if (throwable == null) {
+                        databaseNameConsumer.accept(databaseName);
+                    }
+                })
+                .thenApply(Function.identity());
+    }
+
+    @SuppressWarnings({"DuplicatedCode", "ConstantValue"})
+    public void connect(
```

Comment:
this one is a bit of a beast but not sure if it can broken down to make it more readable. Looks correct though.